src/hotspot/share/gc/shenandoah/heuristics/shenandoahGlobalHeuristics.cpp

```diff
@@ -127,7 +112,8 @@ void ShenandoahGlobalHeuristics::choose_global_collection_set(ShenandoahCollecti
   for (size_t idx = 0; idx < size; idx++) {
     ShenandoahHeapRegion* r = data[idx].get_region();
     if (cset->is_preselected(r->index())) {
-      fatal("There should be no preselected regions during GLOBAL GC");
+      assert(false, "There should be no preselected regions during GLOBAL GC");
```

Comment:
If this is a true invariant (which it is in testing to date), then it must be the case that the call at line 55 above to `add_preselected_regions_to_collection_set()` doesn't do anything, and indeed returns `0` into `cur_young_garbage` there.

Clearly, I am missing something here. @kdnilsen ?

Reply:
I'll run some testing with another assert at the code at line 55 to check my (mis)understanding.

Reply:
You are correct.  In ShenandoahGenerationalHeuristics::choose_collection_set(), we only call prime_collection_set() if generation->is_young().  For global collections, we will not prime the collection set, which means we will not preselect any regions for promotion.

Given this reality, we can remove the call to add_preselected_regions_to_collection_SET() from ShenandoahGlobalHeuristics::choose_collection_set_from_regiondata().  Keeping the assertion in place is how we know it was safe to remove the call to add preselected regions.

We could do some archeology on the code to remind ourselves how we got here.  But maybe not necessary.

Reply:
Could lines 114-118 be replaced with: assert(!cset->is_preselected(r->index(), comment)

Reply:
Thanks Kelvin! I'll convert this to a draft while I do a few experiments/investigations etc. along the lines of your suggestions... (Feel free to leave more comments here as necessary though; thanks!)

Reply:
You are right Kelvin. I made the changes you suggested and will run them through all of the tests. Once that is done, I'll reopen this for a quick re-review. Thanks!

PS: (Archeology) The code was duplicated in https://github.com/openjdk/shenandoah/pull/292 and then the `fatal()` was added in https://github.com/openjdk/shenandoah/pull/318.
Interestingly, for the latter, I had a bunch of review feedback (of a general nature of refactoring) that I never quite got around to submitting, it seems (as I discovered when I loaded the PR in my browser; it's probably pretty stale by now).